cpp/game_main.cpp

```diff
@@ -341,115 +406,97 @@ bool GameMain::on_input(SDL_Event *e) {
-				" UP " << util::FixedPoint<phys_per_tile, 3, 8>{mousepos_phys3.up});
-
-			log::log(MSG(dbg) <<
-				"LMB [tile]:     "
-				" NE " << std::setw(8) << mousepos_tile.ne <<
-				" SE " << std::setw(8) << mousepos_tile.se);
-
-			TerrainChunk *chunk = terrain->get_create_chunk(mousepos_tile);
-			chunk->get_data(mousepos_tile)->terrain_id = editor_current_terrain;
-		}
-		else if (clicking_active and e->button.button == SDL_BUTTON_RIGHT and !construct_mode) {
+			break;
+		case SDL_BUTTON_MIDDLE:
+			if (scrolling_active) { // Stop scrolling
+				SDL_SetRelativeMouseMode(SDL_FALSE);
+				scrolling_active = false;
 
-			// right click can cancel building placement
-			if (this->building_placement) {
-				this->building_placement = false;
-			}
-			else {
-				auto cmd = this->get_action(mousepos_phys3);
-				selection.all_invoke(cmd);
-			}
-		}
-		else if (clicking_active and e->button.button == SDL_BUTTON_RIGHT and construct_mode) {
-			// get chunk clicked on, don't create it if it's not there already
-			// -> placing buildings in void is forbidden that way
-			TerrainChunk *chunk = terrain->get_chunk(mousepos_tile);
-			if (chunk == nullptr) {
-				break;
+				// reactivate mouse clicks as scrolling is over
+				clicking_active = true;
 			}
+			break;
+		case SDL_BUTTON_RIGHT:
+			if (clicking_active) {
+				if (construct_mode) {
+					// get chunk clicked on, don't create it if it's not there already
+					// -> placing buildings in void is forbidden that way
+					TerrainChunk *chunk = terrain->get_chunk(mousepos_tile);
+					if (chunk == nullptr) {
+						break;
+					}
 
-			// delete any unit on the tile
-			if (!chunk->get_data(mousepos_tile)->obj.empty()) {
-				// get first object currently standing at the clicked position
-				TerrainObject *obj = chunk->get_data(mousepos_tile)->obj[0];
-				log::log(MSG(dbg) << "delete unit with unit id " << obj->unit.id);
-				obj->unit.delete_unit();
-			} else if ( this->datamanager.producer_count() > 0 ) {
-				// try creating a unit
-				log::log(MSG(dbg) << "create unit with producer id " << this->editor_current_building);
-				UnitType &producer = *this->datamanager.get_type_index(this->editor_current_building);
-				this->placed_units.new_unit(producer, this->players[rand() % this->players.size()], mousepos_tile.to_phys2().to_phys3());
+					// delete any unit on the tile
+					if (!chunk->get_data(mousepos_tile)->obj.empty()) {
+						// get first object currently standing at the clicked position
+						TerrainObject *obj = chunk->get_data(mousepos_tile)->obj[0];
+						log::log(MSG(dbg) << "delete unit with unit id " << obj->unit.id);
+						obj->unit.delete_unit();
+					} else if ( this->datamanager.producer_count() > 0 ) {
+						// try creating a unit
+						log::log(MSG(dbg) << "create unit with producer id " << this->editor_current_building);
+						UnitType &producer = *this->datamanager.get_type_index(this->editor_current_building);
```

Comment:
Did you just mean the `producer` variable on line 482? Or is something else about it wrong as well? I'm not sure, because I moved these lines without changing them.